core/src/navigation/services/navigation.js

```diff
@@ -87,33 +88,60 @@ class NavigationClass {
             node,
             '_childrenProvider',
             context || node.context
-          )) || [];
-        node.children = node._children.filter(child =>
-          NavigationHelpers.isNodeAccessPermitted(child, node, context)
-        );
-        this.bindChildrenToParent(node);
+          ))
+            .map(n => this.getExpandStructuralPathSegment(n))
+            .map(n => this.bindChildToParent(n, node)) || [];
+
+        node.children = this.getAccessibleNodes(node, context);
         return node.children;
       } catch (err) {
         console.error('Could not lazy-load children for node', err);
       }
     } else if (node._children) {
-      node.children = node._children.filter(child =>
-        NavigationHelpers.isNodeAccessPermitted(child, node, context)
-      );
-      this.bindChildrenToParent(node);
+      node.children = this.getAccessibleNodes(node, context);
       return node.children;
     } else {
       return [];
     }
   }
 
-  bindChildrenToParent(node) {
+  getAccessibleNodes(node, context) {
+    return node._children.filter(child =>
+      NavigationHelpers.isNodeAccessPermitted(child, node, context)
+    );
+  }
+
+  bindChildToParent(child, node) {
+    // Checking for pathSegment to exclude virtual root node
+    // node.pathSegment check is also required for virtual nodes like categories
+    if (node && node.pathSegment) {
+      child.parent = node;
+    }
+    return child;
+  }
+
+  getExpandStructuralPathSegment(node) {
     // Checking for pathSegment to exclude virtual root node
-    if (node && node.pathSegment && node.children) {
-      node.children.forEach(child => {
-        child.parent = node;
-      });
+    if (node && node.pathSegment && node.pathSegment.indexOf('/') !== -1) {
+      const segs = node.pathSegment.split('/');
+      const clonedNode = merge({}, node);
```

Comment:
We can get rid of the 'lodash.merge' dependency here by doing just this: 

`const clonedNode = { ...node };`

Reply:
Unfortunately this is not working.
I need to deep clone properly without keeping references. Try this in the console
```
const node = { label: 'a' };
const context = { key: 'value' };
node.context = context;

const copy = { ...node };
context.key = 'changed!';
// copy context.key should still be 'value' if its a deep clone.
console.log(copy.context.key)
```

Reply:
I understand, but why do we need a deep copy here? 

From

```
{
	pathSegment: 'environments/:environmentId'
	viewUrl: 'sampleapp.html/#envs/:environmentId',
	context: {
		key: 'value'
	}
}
```

we do

```
{
	pathSegment: '/environments'
	children: [
		{
			pathSegment: ':environmentId'
			viewUrl: 'sampleapp.html/#envs/:environmentId',
			context: {
				key: 'value'
			}
		}
	]
}
```

Yes, the context is the same object (and other stuff like children - if the node would have children -) so this is not a pure non-mutable operation but in the current state we reuse the configuration node tree anyway so this is not a big deal. We use the contexts of navigation nodes as they are. In this case we just "attach" context, children etc to the newly created node.   Well, it is not ideal atm but the non-deep copy of node attributes would be closer to the current "architecture" :-). 
In the future we might go with a solution where we do not touch the navigation tree at all and build our own representation based on the configuration but that would be a bigger refactoring.
We can discuss it in an after-scrum.